// eslint-plugin-formatjs/rules/blocklist-elements.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const icu_messageformat_parser_1 = require("@formatjs/icu-messageformat-parser");
const util_1 = require("../util");
class BlacklistElement extends Error {
    constructor(type) {
        super();
        this.message = `${type} element is blocklisted`;
    }
}
var Element;
(function (Element) {
    Element["literal"] = "literal";
    Element["argument"] = "argument";
    Element["number"] = "number";
    Element["date"] = "date";
    Element["time"] = "time";
    Element["select"] = "select";
    Element["selectordinal"] = "selectordinal";
    Element["plural"] = "plural";
    Element["tag"] = "tag";
})(Element || (Element = {}));
function verifyAst(blocklist, ast) {
    for (const el of ast) {
        if ((0, icu_messageformat_parser_1.isLiteralElement)(el) && blocklist.includes(Element.literal)) {
            throw new BlacklistElement(Element.literal);
        }
        if ((0, icu_messageformat_parser_1.isArgumentElement)(el) && blocklist.includes(Element.argument)) {
            throw new BlacklistElement(Element.argument);
        }
        if ((0, icu_messageformat_parser_1.isNumberElement)(el) && blocklist.includes(Element.number)) {
            throw new BlacklistElement(Element.number);
        }
        if ((0, icu_messageformat_parser_1.isDateElement)(el) && blocklist.includes(Element.date)) {
            throw new BlacklistElement(Element.date);
        }
        if ((0, icu_messageformat_parser_1.isTimeElement)(el) && blocklist.includes(Element.time)) {
            throw new BlacklistElement(Element.time);
        }
        if ((0, icu_messageformat_parser_1.isSelectElement)(el) && blocklist.includes(Element.select)) {
            throw new BlacklistElement(Element.select);
        }
        if ((0, icu_messageformat_parser_1.isTagElement)(el) && blocklist.includes(Element.tag)) {
            throw new BlacklistElement(Element.tag);
        }
        if ((0, icu_messageformat_parser_1.isPluralElement)(el)) {
            if (blocklist.includes(Element.plural)) {
                throw new BlacklistElement(Element.argument);
            }
            if (el.pluralType === 'ordinal' &&
                blocklist.includes(Element.selectordinal)) {
                throw new BlacklistElement(Element.selectordinal);
            }
        }
        if ((0, icu_messageformat_parser_1.isSelectElement)(el) || (0, icu_messageformat_parser_1.isPluralElement)(el)) {
            const { options } = el;
            for (const selector of Object.keys(options)) {
                verifyAst(blocklist, options[selector].value);
            }
        }
    }
}
function checkNode(context, node) {
    const settings = (0, util_1.getSettings)(context);
    const msgs = (0, util_1.extractMessages)(node, settings);
    if (!msgs.length) {
        return;
    }
    const blocklist = context.options[0];
    if (!Array.isArray(blocklist) || !blocklist.length) {
        return;
    }
    for (const [{ message: { defaultMessage }, messageNode, },] of msgs) {
        if (!defaultMessage || !messageNode) {
            continue;
        }
        try {
            verifyAst(context.options[0], (0, icu_messageformat_parser_1.parse)(defaultMessage, {
                ignoreTag: settings.ignoreTag,
            }));
        }
        catch (e) {
            context.report({
                node: messageNode,
                message: e instanceof Error ? e.message : String(e),
            });
        }
    }
}
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Disallow specific elements in ICU message format',
            category: 'Errors',
            recommended: false,
            url: 'https://formatjs.io/docs/tooling/linter#blocklist-elements',
        },
        fixable: 'code',
        schema: [
            {
                type: 'array',
                properties: {
                    items: {
                        type: 'string',
                        enum: Object.keys(Element),
                    },
                },
            },
        ],
    },
    create(context) {
        const callExpressionVisitor = (node) => checkNode(context, node);
        if (context.parserServices.defineTemplateBodyVisitor) {
            return context.parserServices.defineTemplateBodyVisitor({
                CallExpression: callExpressionVisitor,
            }, {
                CallExpression: callExpressionVisitor,
            });
        }
        return {
            JSXOpeningElement: (node) => checkNode(context, node),
            CallExpression: callExpressionVisitor,
        };
    },
};
exports.default = rule;
