// eslint-plugin-formatjs/rules/enforce-default-message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("../util");
function checkNode(context, node) {
    const msgs = (0, util_1.extractMessages)(node, (0, util_1.getSettings)(context));
    const { options: [type], } = context;
    for (const [{ message: { defaultMessage }, messageNode, },] of msgs) {
        if (!defaultMessage) {
            if (type === 'literal' && messageNode) {
                context.report({
                    node: messageNode,
                    message: `"defaultMessage" must be:
- a string literal or
- template literal without variable`,
                });
            }
            else if (!messageNode) {
                context.report({
                    node: node,
                    message: '`defaultMessage` has to be specified in message descriptor',
                });
            }
        }
    }
}
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Enforce defaultMessage in message descriptor',
            category: 'Errors',
            recommended: false,
            url: 'https://formatjs.io/docs/tooling/linter#enforce-default-message',
        },
        fixable: 'code',
        schema: [
            {
                enum: ['literal', 'anything'],
            },
        ],
    },
    create(context) {
        const callExpressionVisitor = (node) => checkNode(context, node);
        if (context.parserServices.defineTemplateBodyVisitor) {
            return context.parserServices.defineTemplateBodyVisitor({
                CallExpression: callExpressionVisitor,
            }, {
                CallExpression: callExpressionVisitor,
            });
        }
        return {
            JSXOpeningElement: (node) => checkNode(context, node),
            CallExpression: callExpressionVisitor,
        };
    },
};
exports.default = rule;
