// eslint-plugin-formatjs/rules/no-multiple-plurals.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const icu_messageformat_parser_1 = require("@formatjs/icu-messageformat-parser");
const util_1 = require("../util");
class MultiplePlurals extends Error {
    constructor() {
        super(...arguments);
        this.message = 'Cannot specify more than 1 plural rules';
    }
}
function verifyAst(ast, pluralCount = { count: 0 }) {
    for (const el of ast) {
        if ((0, icu_messageformat_parser_1.isPluralElement)(el)) {
            pluralCount.count++;
            if (pluralCount.count > 1) {
                throw new MultiplePlurals();
            }
            const { options } = el;
            for (const selector of Object.keys(options)) {
                verifyAst(options[selector].value, pluralCount);
            }
        }
    }
}
function checkNode(context, node) {
    const settings = (0, util_1.getSettings)(context);
    const msgs = (0, util_1.extractMessages)(node, settings);
    for (const [{ message: { defaultMessage }, messageNode, },] of msgs) {
        if (!defaultMessage || !messageNode) {
            continue;
        }
        try {
            verifyAst((0, icu_messageformat_parser_1.parse)(defaultMessage, {
                ignoreTag: settings.ignoreTag,
            }));
        }
        catch (e) {
            context.report({
                node: messageNode,
                message: e instanceof Error ? e.message : String(e),
            });
        }
    }
}
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Disallow multiple plural rules in the same message',
            category: 'Errors',
            recommended: false,
            url: 'https://formatjs.io/docs/tooling/linter#no-multiple-plurals',
        },
        fixable: 'code',
    },
    create(context) {
        const callExpressionVisitor = (node) => checkNode(context, node);
        if (context.parserServices.defineTemplateBodyVisitor) {
            return context.parserServices.defineTemplateBodyVisitor({
                CallExpression: callExpressionVisitor,
            }, {
                CallExpression: callExpressionVisitor,
            });
        }
        return {
            JSXOpeningElement: (node) => checkNode(context, node),
            CallExpression: callExpressionVisitor,
        };
    },
};
exports.default = rule;
