// eslint-plugin-formatjs/rules/no-useless-message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const icu_messageformat_parser_1 = require("@formatjs/icu-messageformat-parser");
const util_1 = require("../util");
class JustArgument extends Error {
    constructor() {
        super(...arguments);
        this.message = 'Unnecessary formatted message.';
    }
}
class JustNumber extends Error {
    constructor() {
        super(...arguments);
        this.message = 'Unnecessary formatted message: just use FormattedNumber or intl.formatNumber.';
    }
}
class JustDate extends Error {
    constructor() {
        super(...arguments);
        this.message = 'Unnecessary formatted message: just use FormattedDate or intl.formatDate.';
    }
}
class JustTime extends Error {
    constructor() {
        super(...arguments);
        this.message = 'Unnecessary formatted message: just use FormattedTime or intl.formatTime.';
    }
}
function verifyAst(ast) {
    if (ast.length !== 1) {
        return;
    }
    switch (ast[0].type) {
        case icu_messageformat_parser_1.TYPE.argument:
            throw new JustArgument();
        case icu_messageformat_parser_1.TYPE.number:
            throw new JustNumber();
        case icu_messageformat_parser_1.TYPE.date:
            throw new JustDate();
        case icu_messageformat_parser_1.TYPE.time:
            throw new JustTime();
    }
}
function checkNode(context, node) {
    const settings = (0, util_1.getSettings)(context);
    const msgs = (0, util_1.extractMessages)(node, settings);
    for (const [{ message: { defaultMessage }, messageNode, },] of msgs) {
        if (!defaultMessage || !messageNode) {
            continue;
        }
        try {
            verifyAst((0, icu_messageformat_parser_1.parse)(defaultMessage, {
                ignoreTag: settings.ignoreTag,
            }));
        }
        catch (e) {
            context.report({
                node: messageNode,
                message: e instanceof Error ? e.message : String(e),
            });
        }
    }
}
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Disallow unnecessary formatted message',
            recommended: true,
            url: 'https://formatjs.io/docs/tooling/linter#no-useless-message',
        },
        fixable: 'code',
    },
    create(context) {
        const callExpressionVisitor = (node) => checkNode(context, node);
        if (context.parserServices.defineTemplateBodyVisitor) {
            return context.parserServices.defineTemplateBodyVisitor({
                CallExpression: callExpressionVisitor,
            }, {
                CallExpression: callExpressionVisitor,
            });
        }
        return {
            JSXOpeningElement: (node) => checkNode(context, node),
            CallExpression: callExpressionVisitor,
        };
    },
};
exports.default = rule;
