// @cloudbeds/webpack-module-federation-types-plugin/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CLOUDBEDS_DEPLOYMENT_ENV_WITH_DISABLED_REMOTE_TYPES_DOWNLOAD = exports.CLOUDBEDS_REMOTES_MANIFEST_FILE_NAME = exports.CloudbedsCloudfrontDomain = exports.DEFAULT_DOWNLOAD_TYPES_INTERVAL_IN_SECONDS = exports.DEFAULT_DIR_DOWNLOADED_TYPES = exports.DEFAULT_DIR_GLOBAL_TYPES = exports.DEFAULT_DIR_EMITTED_TYPES = exports.DEFAULT_DIR_DIST = exports.TS_CONFIG_FILE = exports.FEDERATION_CONFIG_FILE = void 0;
exports.FEDERATION_CONFIG_FILE = 'federation.config.json';
exports.TS_CONFIG_FILE = 'tsconfig.json';
exports.DEFAULT_DIR_DIST = 'dist';
exports.DEFAULT_DIR_EMITTED_TYPES = '@types';
exports.DEFAULT_DIR_GLOBAL_TYPES = 'src/@types';
exports.DEFAULT_DIR_DOWNLOADED_TYPES = `src/@types/remotes`;
exports.DEFAULT_DOWNLOAD_TYPES_INTERVAL_IN_SECONDS = 60;
var CloudbedsCloudfrontDomain;
(function (CloudbedsCloudfrontDomain) {
    CloudbedsCloudfrontDomain["Dev"] = "https://cb-front.cloudbeds-dev.com";
    CloudbedsCloudfrontDomain["Stage"] = "https://cb-front.cloudbeds-stage.com";
    CloudbedsCloudfrontDomain["Prod"] = "https://front.cloudbeds.com";
})(CloudbedsCloudfrontDomain = exports.CloudbedsCloudfrontDomain || (exports.CloudbedsCloudfrontDomain = {}));
exports.CLOUDBEDS_REMOTES_MANIFEST_FILE_NAME = 'remote-entries.json';
exports.CLOUDBEDS_DEPLOYMENT_ENV_WITH_DISABLED_REMOTE_TYPES_DOWNLOAD = 'devbox';
